From 6ec6f411a23762f85de4de3a92cb03b2709d4f5a Mon Sep 17 00:00:00 2001
From: Hui Min Mina Chou <minachou@andestech.com>
Date: Mon, 26 Aug 2024 10:13:40 +0800
Subject: [PATCH 156/172] lib: utils: cache: L2C init is only needed at cold
 boot (#117)

L2C init is required only once during cold boot. Commit d5f1ca13c458
("lib: utils: cache: Add L2C root domain region with R/W permissions for
both M/S-mode.") caused it to run during warm boot, triggering the error:
"fdt_cache_driver_init: cache init failed, -3". This happens because,
after cold boot, the root domain region is protected [1]. Thus, when warm
boot tries to initialize L2C and attempts to add the root domain region
[2], it fails.

[1]: https://github.com/riscv-software-src/opensbi/blob/43cace6c3671e5172d0df0a8963e552bb04b7b20/lib/sbi/sbi_domain.c#L755
[2]: https://gitea.andestech.com/RD-SW/opensbi/src/commit/d5f1ca13c45847e6e1e203a50c0b532575c1197e/lib/utils/cache/fdt_cache_andes_l2c.c#L178

Fixes: d5f1ca13c458 ("lib: utils: cache: Add L2C root domain region with R/W permissions for both M/S-mode.")
Signed-off-by: Hui Min Mina Chou <minachou@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/opensbi/pulls/117
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: Ben Zong-You Xie <ben717@andestech.com>
Reviewed-by: Peter Yu-Chien Lin <peterlin@andestech.com>
Reviewed-by: randolph <randolph@andestech.com>
Co-authored-by: Hui Min Mina Chou <minachou@andestech.com>
Co-committed-by: Hui Min Mina Chou <minachou@andestech.com>
---
 platform/generic/andes/ae350.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/platform/generic/andes/ae350.c b/platform/generic/andes/ae350.c
index 9a6e771b7d12..f41c8c03d1ca 100644
--- a/platform/generic/andes/ae350.c
+++ b/platform/generic/andes/ae350.c
@@ -120,7 +120,9 @@ static int ae350_final_init(bool cold_boot, const struct fdt_match *match)
 
 static int ae350_early_init(bool cold_boot, const struct fdt_match *match)
 {
-	fdt_cache_init();
+	if (cold_boot)
+		fdt_cache_init();
+
 	return 0;
 }
 
-- 
2.34.1

